### ISSUE TITLE
JavaFX Threads with DL4J won't work - Same score all the time

### ISSUE THREAD

**Issue by DanielMartensson** [deeplearning4j/deeplearning4j#8343]
#### Issue Description

I'm working on something called Deeplearning2C. It's a C-code generator that generate C-code with a BLAS library. https://github.com/DanielMartensson/Deeplearning2C

When I train my model, then I get the same score all the time. But when I restart the training, then I get only ONE score less than before. So I assume that threads + DL4J won't work very well together.

When I train my model like this:

```
                       // Train
			try {
				multiLayerNetwork.fit(trainDataSetIterator);
			}catch(DL4JException | IllegalStateException e) {
				Platform.runLater(() -> textArea.setText(e.getMessage())); 
				endMessage = "\nStatus: Training fail...";
				break;
			}
```
I get this output when this statement is running: `multiLayerNetwork.fit(trainDataSetIterator);`

> 21:08:59.473 [Thread-1333] DEBUG org.nd4j.linalg.dataset.AsyncDataSetIterator - Manually destroying ADSI workspace
> 21:08:59.474 [Thread-1333] DEBUG org.nd4j.linalg.dataset.AsyncDataSetIterator - Manually destroying ADSI workspace

Just download Deeplarning2c and run it with `./gradlew run` for Linux or `gradlew run` for Windows. Then you need to add some data and configure the layers. I recommend the iris.txt data below inside my Readme.md file of the DL2C repository. 

Try to train and then you will see that the score won't update, unless you restart the training again. But the score only get one step less, than before. So I assume that this is a thread issue for DL4J.

#### Version Information

Please indicate relevant versions, including, if relevant:

* Deeplearning4j version = 1.0.0-Beta5
* Platform information (OS, etc) = Ubuntu Linux 18.04
* CUDA version, if used = No
* NVIDIA driver version, if in use = No Nvidia. Just plain simple DL4J.

#### Additional Information

Here is my graldew file.

```
buildscript {
    repositories {
        jcenter()
    }
    dependencies {
        classpath 'org.javafxports:jfxmobile-plugin:1.3.17'
    }
}

apply plugin: 'org.javafxports.jfxmobile'

repositories {
    jcenter()
    maven {
        url 'http://nexus.gluonhq.com/nexus/content/repositories/releases'
    }
}

mainClassName = 'se.danielmartensson.deeplearning2c.Main'

configurations {
     all*.exclude module: 'hdf5'
     all*.exclude module: 'hdf5-platform'
     all*.exclude module: 'leptonica'
     all*.exclude module: 'opencv'
     all*.exclude module: 'opencsv'
     all*.exclude module: 'datavec-data-image'
     all*.exclude module: 'deeplearning4j-ui-components'
     all*.exclude module: 'joda-time'
     all*.exclude module: 'gson'
     all*.exclude module: 'sneakeyaml'
     all*.exclude module: 'threetenbp'
     all*.exclude module: 'stax2-api'
     all*.exclude module: 'commons-net'
     all*.exclude module: 'neoitertools'

}

dependencies {
    compile 'com.gluonhq:charm:5.0.2'
    compile group: 'org.deeplearning4j', name: 'deeplearning4j-core', version: '1.0.0-beta5'
    compile group: 'org.nd4j', name: 'nd4j-native-platform', version: '1.0.0-beta5'
    compile group: 'org.datavec', name: 'datavec-api', version: '1.0.0-beta5'
    compile group: 'org.projectlombok', name: 'lombok', version: '1.18.8'
    compile group: 'ch.qos.logback', name: 'logback-classic', version: '1.2.3'
}

jfxmobile {
    downConfig {
        version = '3.8.6'
        // Do not edit the line below. Use Gluon Mobile Settings in your project context menu instead
        plugins 'display', 'lifecycle', 'statusbar', 'storage'
    }
    android {
        manifest = 'src/android/AndroidManifest.xml'
        targetSdkVersion = '23'
        dexOptions {
            additionalParameters = ['--minimal-main-dex']
            keepRuntimeAnnotatedClasses = false
            javaMaxHeapSize = '4g'
        }
        packagingOptions {
            exclude 'META-INF/DEPENDENCIES'
            exclude 'META-INF/DEPENDENCIES.txt'
            exclude 'META-INF/LICENSE'
            exclude 'META-INF/LICENSE.txt'
            exclude 'LICENSE-EDL-1.0.txt'
            exclude 'LICENSE-EPL-1.0.txt'
            exclude 'about.html'
            exclude 'META-INF/license.txt'
            exclude 'META-INF/NOTICE'
            exclude 'META-INF/NOTICE.txt'
            exclude 'META-INF/notice.txt'
            exclude 'META-INF/INDEX.LIST'
            exclude 'META-INF/services/javax.imageio.spi.ImageReaderSpi'
            exclude 'META-INF/services/javax.imageio.spi.ImageWriterSpi'
            exclude 'assets/org/apache/commons/math3/exception/util/LocalizedFormats_fr.properties'
            exclude 'assets/org/apache/commons/math3/random/new-joe-kuo-6.1000'
            exclude 'org/threeten/bp/format/ChronologyText.properties'
            exclude 'org/threeten/bp/TZDB.dat'
            pickFirst 'META-INF/services/org.nd4j.linalg.factory.Nd4jBackend'
            pickFirst 'META-INF/services/org.nd4j.linalg.compression.NDArrayCompressor'
            pickFirst 'nd4j-native.properties'
        }
    }
    ios {
        infoPList = file('src/ios/Default-Info.plist')
        forceLinkClasses = [
                'se.danielmartensson.**.*',
                'com.gluonhq.**.*',
                'javax.annotations.**.*',
                'javax.inject.**.*',
                'javax.json.**.*',
                'org.glassfish.json.**.*'
        ]
    }
}
```

#### Contributing

Here is the thread I'm using.

```
package se.danielmartensson.deeplearning2c.deeplearning;


import java.util.concurrent.atomic.AtomicBoolean;

import org.deeplearning4j.exception.DL4JException;
import org.deeplearning4j.nn.multilayer.MultiLayerNetwork;
import org.nd4j.linalg.dataset.api.iterator.DataSetIterator;

import com.gluonhq.charm.glisten.control.AppBar;
import com.gluonhq.charm.glisten.control.ProgressBar;

import javafx.application.Platform;
import javafx.scene.control.Button;
import javafx.scene.control.TextArea;


public class DL4JThread extends Thread{
	private MultiLayerNetwork multiLayerNetwork;
	ProgressBar progressBar;
	private AtomicBoolean continueLoop;
	private int epochs;
	private DataSetIterator trainDataSetIterator;
	private TextArea textArea;
	private int progressBarPosition;
	private String endMessage;
	private String[] textWall = new String[31]; // 31 elements
	private AppBar appBar;
	private Button startButton;
	
	/**
	 * Constructor for model fitting
	 * @param multiLayerNetwork
	 * @param trainDataSetIterator
	 * @param textArea
	 * @param progressBar
	 * @param continueLoop
	 * @param epochs
	 * @param startButton 
	 * @param appBar 
	 */
	public DL4JThread(MultiLayerNetwork multiLayerNetwork, DataSetIterator trainDataSetIterator, TextArea textArea, ProgressBar progressBar, AtomicBoolean continueLoop, int epochs, AppBar appBar, Button startButton) {
		this.multiLayerNetwork = multiLayerNetwork;
		this.trainDataSetIterator = trainDataSetIterator;
		this.textArea = textArea;
		this.progressBar = progressBar;
		this.continueLoop = continueLoop;
		this.epochs = epochs;
		this.appBar = appBar;
		this.startButton = startButton;
	}

	@Override
	public void run() {
		// Initial counters
		progressBarPosition = 1;
		int rowCounter = 0;
		endMessage = "\nStatus: Training done...";
		
		// While loop for training
		while(progressBarPosition < epochs+1) {
			// If we press stop or let the for loop do its full iterations
			if(continueLoop.get() == false)
				break; 
			
			// Train
			try {
				multiLayerNetwork.fit(trainDataSetIterator);
			}catch(DL4JException | IllegalStateException e) {
				Platform.runLater(() -> textArea.setText(e.getMessage())); 
				endMessage = "\nStatus: Training fail...";
				break;
			}
			
			// Delay is needed, else we might lose some icons inside the appBar
			try {
				Thread.sleep(20);   
			} catch (Exception e) {
				e.getStackTrace();
			}
			
			// Print the score out 
			if(rowCounter >= textWall.length) {
				// Get the score and add it to the last element
				String rowLine = "Epoch = " + progressBarPosition + " Score = " + multiLayerNetwork.score();
				for(int j = 0; j < textWall.length-1; j++) 
					textWall[j] = textWall[j+1]; // Shift
				textWall[textWall.length-1] = rowLine; // Last
				Platform.runLater(() -> textArea.setText(String.join("\n", textWall))); 
			}else {
				String rowLine = "Epoch = " + progressBarPosition + " Score = " + multiLayerNetwork.score();
				textWall[rowCounter] = rowLine;
				Platform.runLater(() -> textArea.appendText(rowLine + "\n")); 
				rowCounter++;
			}	
			
			// Our progress bar
			Platform.runLater(() -> progressBar.setProgress(Double.valueOf(progressBarPosition)/Double.valueOf(epochs)));
			
			// Break statement 
			if(progressBarPosition >= epochs)
				break;
			else
				progressBarPosition++; // I use a while-loop due to the Platform.RunLater cannot take variable i as a local variable.
		}
		
		// Print the status message
		Platform.runLater(() -> textArea.appendText(endMessage));
		
		// Change the icon to start button
		Platform.runLater(() -> {
			appBar.getActionItems().remove(0);
			appBar.getActionItems().add(0, startButton);
		});
	}
}

```

![a](https://i.ibb.co/PtbXBnP/Selecci-n-022.png)

**Comment by saudet** [comment]
Are you able to reproduce that without JavaFX?

**Comment by DanielMartensson** [comment]
> Are you able to reproduce that without JavaFX?

No. I dont think so. This happens when I use a thread. If I restart the thread then the number score goes down one step. So I assume that DL4J is not thread safe. 

It worked before with 1.0.0-Beta4. But now it won't work at all for both 1.0.0-Beta4 and Beta5.

**Comment by raver119** [comment]
DL4J is definitely NOT thread safe. With the exception of classes, which are thread safe by design. Such classes has thread safety mentioned in docs.

**Comment by DanielMartensson** [comment]
> DL4J is definitely NOT thread safe. With the exception of classes, which are thread safe by design. Such classes has thread safety mentioned in docs.

So. Do you recommend to not use a thread in this case? I'm only use a thread for the update of JavaFX components.

**Comment by raver119** [comment]
If you access 1 model from 1 thread all the time - it has nothing to do with thread safety. Problems will arise once you'll try to access model from multiple threads at the same time.

**Comment by DanielMartensson** [comment]
> If you access 1 model from 1 thread all the time - it has nothing to do with thread safety. Problems will arise once you'll try to access model from multiple threads at the same time.

Well, I have only one thread to use with the model. So this is a JavaFX issue?

**Comment by raver119** [comment]
trainDataSetIterator <--- are you sure it's properly reset before epoch starts?

**Comment by DanielMartensson** [comment]
> trainDataSetIterator <--- are you sure it's properly reset before epoch starts?

I don't reset it. I don't know how. I know that some functions are obselete now.

**Comment by raver119** [comment]
> trainDataSetIterator.reset();

Add this after model.fit() call

**Comment by raver119** [comment]
Also, can you please add something like ScoreIterationListener to your model, just to see what actually happens during your epochs?

**Comment by DanielMartensson** [comment]
> Also, can you please add something like ScoreIterationListener to your model, just to see what actually happens during your epochs?

But I'm printing the score?

**Comment by DanielMartensson** [comment]
> > trainDataSetIterator.reset();
> 
> Add this after model.fit() call

I add this code

  

```
			try {
				multiLayerNetwork.fit(trainDataSetIterator);
				trainDataSetIterator.reset();
			}catch(DL4JException | IllegalStateException e) {
				Platform.runLater(() -> textArea.setText(e.getMessage())); 
				endMessage = "\nStatus: Training fail...";
				break;
			}

```


Result


![image](https://user-images.githubusercontent.com/28505229/68145034-2627c180-ff35-11e9-8196-af8ea44b06ea.png)

When I add `multiLayerNetwork.setListeners(new ScoreIterationListener(1));` to the model.

![image](https://i.ibb.co/XkzbPXX/Selecci-n-024.png)

**Comment by DanielMartensson** [comment]
Give me a reply if you have read this post above. I know that you have much work to do, but I wonder if you have notice the information.

**Comment by raver119** [comment]
So, according to your log there's 2 things happening:
1) you have exactly 1 dataset in your iterator
2) subsequent epochs do not trigger actual training.

Please show how you create your iterator, and source code for it, if it's yours.

**Comment by DanielMartensson** [comment]
> So, according to your log there's 2 things happening:
> 
>     1. you have exactly 1 dataset in your iterator
> 
>     2. subsequent epochs do not trigger actual training.
> 
> 
> Please show how you create your iterator, and source code for it, if it's yours.

Yes! Here!
https://github.com/DanielMartensson/Deeplearning2C/blob/master/Deeplearning2CApp/src/main/java/se/danielmartensson/deeplearning2c/deeplearning/DL4JData.java

**Comment by AlexDBlack** [comment]
@DanielMartensson this looks wrong: you're using the test set twice:
https://github.com/DanielMartensson/Deeplearning2C/blob/master/Deeplearning2CApp/src/main/java/se/danielmartensson/deeplearning2c/deeplearning/DL4JData.java#L50-L53

Also separately (line 39) you're normalizing on both the train and test set, which is considered bad practice. You should collect normalization statistics using the training set only (i.e., DataNormalizer.fit), and use the training set statistics to normalize both the train and test set.

Try again after fixing the bug at line 52/53. And better yet, use a simpler data pipeline (no DataSetIteratorSplitter)... I suspect that's your problem.

**Comment by DanielMartensson** [comment]
> @DanielMartensson this looks wrong: you're using the test set twice:
> https://github.com/DanielMartensson/Deeplearning2C/blob/master/Deeplearning2CApp/src/main/java/se/danielmartensson/deeplearning2c/deeplearning/DL4JData.java#L50-L53
> 
> Also separately (line 39) you're normalizing on both the train and test set, which is considered bad practice. You should collect normalization statistics using the training set only (i.e., DataNormalizer.fit), and use the training set statistics to normalize both the train and test set.
> 
> Try again after fixing the bug at line 52/53. And better yet, use a simpler data pipeline (no DataSetIteratorSplitter)... I suspect that's your problem.

Hi!
Yes! You solved the problem. It was the test set I used twice.

Wait! So you mean that scaling the training data between intervals (min, max) for training by using NormalizerMinMaxScaler class is bad practice. I should use  DataNormalizer class instead? I'm only want to scale my data to -1 and 1, or 0 to 1 etc. Just for the activation functions. 

Simpler data pipeline? What type of pipeline?

**Comment by AlexDBlack** [comment]
> So you mean that scaling the training data between intervals (min, max) for training by using NormalizerMinMaxScaler class is bad practice. I should use DataNormalizer class instead?

DataNormalizer is an interface. Using NormalizerMinMaxScalar is fine.
But you want to use it on the training set only.
So do train/test split, then call fit with training set only, not the entire dataset.
Otherwise you're leaking a small amount of information from test set into training - so your test set is no longer "unseen" if that makes sense.


> Simpler data pipeline?

For splitting (from something like a CSV) I prefer to do it as a separate step.
So load your data, split it, save the test set and training sets separately. Then load them up separately.

**Comment by DanielMartensson** [comment]
> > So you mean that scaling the training data between intervals (min, max) for training by using NormalizerMinMaxScaler class is bad practice. I should use DataNormalizer class instead?
> 
> DataNormalizer is an interface. Using NormalizerMinMaxScalar is fine.
> But you want to use it on the training set only.
> So do train/test split, then call fit with training set only, not the entire dataset.
> Otherwise you're leaking a small amount of information from test set into training - so your test set is no longer "unseen" if that makes sense.
> 
> > Simpler data pipeline?
> 
> For splitting (from something like a CSV) I prefer to do it as a separate step.
> So load your data, split it, save the test set and training sets separately. Then load them up separately.

Like this:

```
public void loadData(File fileCSV, char delimiter, int batchSize, int labelIndexFrom, int labelIndexTo, int numPossibleLabels, boolean regression, int minScalar, int maxScalar) throws IOException, InterruptedException {
		// Get the all data sets
		dataSetIterator = readCSVDataset(fileCSV, delimiter, batchSize, labelIndexFrom, labelIndexTo, numPossibleLabels, regression);
		
		// Calculate totalBatches = amount of data sets X batch size for each data set
		long totalBatches = 0;
		while(dataSetIterator.hasNext() == true) {
			dataSetIterator.next();
			totalBatches++;
		}
		dataSetIterator.reset();
		totalBatches = totalBatches*dataSetIterator.batch();
		
		// Split the data into 50% train and 50% eval
		DataSetIteratorSplitter dataSetIteratorSplitter = new DataSetIteratorSplitter(dataSetIterator, totalBatches, 0.5);
		trainDataSetIterator = dataSetIteratorSplitter.getTrainIterator();
		evalDataSetIterator = dataSetIteratorSplitter.getTestIterator();
	
		// Normalize the train data iterators
		NormalizerMinMaxScaler normalizerMinMaxScaler = new NormalizerMinMaxScaler(minScalar, maxScalar);
		normalizerMinMaxScaler.fit(trainDataSetIterator);

	}
```

I think I need to fit the test data as well because I'm need of a test data in a specific range. Because the data have large numbers. Both negative and positive.
